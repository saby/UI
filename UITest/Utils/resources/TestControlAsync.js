define('UITest/Utils/resources/TestControlAsync', [
    'UI/Base',
    'wml!UITest/Utils/resources/TestControlAsync'
],
function (UiBase, template) {
    return {
        default: UiBase.Control.extend({
            _template: template
        })
    };
});
