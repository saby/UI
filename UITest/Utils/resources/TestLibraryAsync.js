define(['tslib', 'UI/Base', 'wml!UITest/Utils/resources/TestControlAsync'], function (tslib, Base, template) {
    var ExportControl = /** @class */ (function (_super) {
        tslib.__extends(ExportControl, _super);
        function ExportControl() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this._template = template;
            return _this;
        }
        return ExportControl
    })(Base.Control)

    return {
        exportFunction: function(echo) { return echo; },
        exportSyncFunction: function(echo) { return echo; },
        ExportControl: ExportControl
    };
});
