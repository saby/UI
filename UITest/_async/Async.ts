import { Async, TAsyncStateReceived, IAsyncOptions } from 'UICore/Async';

/**
 * Реализация класса UICore/Async:Async для тестов
 */
export default class AsyncTest extends Async {
    _beforeMount(options: IAsyncOptions): void {
        super._beforeMount(options);
    }

    _componentDidMount() {
        super._componentDidMount();
    }

    _beforeUpdate(opts: IAsyncOptions) {
        super._beforeUpdate(opts);
    }

    _afterUpdate() {
        super._afterUpdate();
    }

    getError(): TAsyncStateReceived | void {
        return this.error;
    }

    getCurrentTemplateName(): string {
        return this.currentTemplateName;
    }

    getOptionsForComponent(): Record<string, unknown> {
        return this.optionsForComponent;
    }
}
