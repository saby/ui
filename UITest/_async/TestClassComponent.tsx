import { Component } from 'react';

export default class TestClassComponent extends Component {
    render() {
        return <div>TestClassComponent</div>;
    }
}
