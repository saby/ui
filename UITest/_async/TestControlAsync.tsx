export default function TestControlAsync() {
    return <div>TestControlAsync</div>;
}
