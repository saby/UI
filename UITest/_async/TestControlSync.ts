import { Control, TemplateFunction } from 'UI/Base';
import template = require('wml!UITest/_async/TestControlSync');

class TestControlSync extends Control {
    protected _template: TemplateFunction = template;
}

export = TestControlSync;
