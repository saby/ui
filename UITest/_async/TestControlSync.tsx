export default function TestControlSync() {
    return <div>TestControlSync</div>;
}
