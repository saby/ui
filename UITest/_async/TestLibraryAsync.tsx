function ExportControl() {
    return <div>ExportControl</div>;
}

export { ExportControl };
