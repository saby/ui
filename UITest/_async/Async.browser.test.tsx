/**
 * @jest-environment jsdom
 */
import { render } from 'react-dom';
import { act } from 'react-dom/test-utils';
import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import { Logger } from 'UICommon/Utils';
import { Async } from 'UI/Async';

// предзагрузим модули, на которых будем тестировать Async - в тестах главное протестировать факт вызова require
require('UITest/_async/TestControlSync');
import TestControlAsync = require('UITest/_async/TestControlAsync');
import TestLibraryAsync = require('UITest/_async/TestLibraryAsync');

describe('UI/Async:Async в браузере', () => {
    let container: HTMLDivElement;

    async function microTaskWait() {
        return Promise.resolve();
    }

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);

        jest.spyOn(Logger, 'error').mockImplementation(jest.fn());
    });

    afterEach(() => {
        jest.useRealTimers();
        container.remove();

        jest.restoreAllMocks();
    });

    it('Синхронная загрузка контрола', () => {
        const moduleName = 'UITest/_async/TestControlSync';

        act(() => {
            render(<Async componentName={moduleName} componentProps={{}} />, container);
        });

        expect(container).toMatchSnapshot();
    });

    it('Вызов коллбека onComponentLoad после синхронной загрузки контрола', () => {
        const moduleName = 'UITest/_async/TestControlSync';
        const onComponentLoad = jest.fn();

        act(() => {
            render(
                <Async
                    componentName={moduleName}
                    componentProps={{}}
                    onComponentLoad={onComponentLoad}
                />,
                container
            );
        });

        expect(onComponentLoad).toBeCalledTimes(1);
    });

    it('Синхронная загрузка контрола, с ошибкой', () => {
        const moduleName = 'UIdemo/_async/Fail/TestControlSync';

        // заглушка для проверки факта вызова загрузки в "require"
        const loadSyncSpy = jest.spyOn(ModulesLoader, 'loadSync').mockReturnValueOnce(null);
        // заглушка для ModulesLoader.isLoaded, чтобы проверить синхронную загрузку контрола
        jest.spyOn(ModulesLoader, 'isLoaded').mockReturnValue(true);

        act(() => {
            render(<Async componentName={moduleName} componentProps={{}} />, container);
        });

        // проверим, что была попытка синхронной загрузки контрола
        expect(loadSyncSpy).toBeCalled();
        expect(loadSyncSpy).toBeCalledWith(moduleName);
        expect(container).toMatchSnapshot();
    });

    it('Асинхронная загрузка контрола', async () => {
        const moduleName = 'UITest/_async/TestControlAsync';
        // заглушка для проверки факта вызова загрузки в "require"
        const loadAsyncSpy = jest
            .spyOn(ModulesLoader, 'loadAsync')
            .mockResolvedValueOnce(TestControlAsync);
        // заглушка для ModulesLoader.isLoaded, чтобы проверить асинхронную загрузку контрола
        jest.spyOn(ModulesLoader, 'isLoaded').mockReturnValue(false);
        const onComponentLoad = jest.fn();

        act(() => {
            render(
                <Async
                    componentName={moduleName}
                    componentProps={{}}
                    onComponentLoad={onComponentLoad}
                />,
                container
            );
        });

        // проверим, что был вызов асинхронной загрузки контрола
        expect(loadAsyncSpy).toBeCalled();
        expect(loadAsyncSpy).toBeCalledWith(moduleName);

        await microTaskWait();

        // Должен быть вызван callback onComponentLoad
        expect(onComponentLoad).toBeCalledTimes(1);

        expect(container).toMatchSnapshot();
    });

    it('Асинхронная загрузка контрола, с ошибкой', async () => {
        const moduleName = 'UITest/_async/Fail/TestControlAsync';
        // заглушка для проверки факта вызова загрузки в "require"
        const loadAsyncSpy = jest.spyOn(ModulesLoader, 'loadAsync').mockRejectedValueOnce(null);
        // мокаем, чтобы не было сообщения в консоли
        jest.spyOn(ModulesLoader, 'unloadSync').mockImplementationOnce(jest.fn());
        const onComponentLoad = jest.fn();

        act(() => {
            render(
                <Async
                    componentName={moduleName}
                    componentProps={{}}
                    onComponentLoad={onComponentLoad}
                />,
                container
            );
        });

        // проверим, что был вызов асинхронной загрузки контрола
        expect(loadAsyncSpy).toBeCalled();
        expect(loadAsyncSpy).toBeCalledWith(moduleName);

        await microTaskWait();

        expect(onComponentLoad).not.toBeCalled(); // не должно быть вызова callback'а onComponentLoad
        expect(container).toMatchSnapshot();
    });

    it('Асинхронная загрузка из библиотеки', async () => {
        const moduleName = 'UITest/_async/TestLibraryAsync:ExportControl';
        // заглушка для проверки факта вызова загрузки в "require"
        const loadAsyncSpy = jest
            .spyOn(ModulesLoader, 'loadAsync')
            .mockResolvedValueOnce(TestLibraryAsync.ExportControl);
        // заглушка для ModulesLoader.isLoaded, чтобы проверить асинхронную загрузку контрола
        jest.spyOn(ModulesLoader, 'isLoaded').mockReturnValue(false);

        act(() => {
            render(<Async componentName={moduleName} componentProps={{}} />, container);
        });

        // проверим, что был вызов асинхронной загрузки контрола
        expect(loadAsyncSpy).toBeCalled();
        expect(loadAsyncSpy).toBeCalledWith(moduleName);

        await microTaskWait();

        expect(container).toMatchSnapshot();
    });
});
