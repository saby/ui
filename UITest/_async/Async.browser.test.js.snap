// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Async:Async в браузере Асинхронная загрузка из библиотеки 1`] = `
<div>
  <div>
    ExportControl
  </div>
</div>
`;

exports[`UI/Async:Async в браузере Асинхронная загрузка контрола 1`] = `
<div>
  <div>
    TestControlAsync
  </div>
</div>
`;

exports[`UI/Async:Async в браузере Асинхронная загрузка контрола, с ошибкой 1`] = `
<div>
  <div>
    У СБИС возникла проблема
  </div>
</div>
`;

exports[`UI/Async:Async в браузере Синхронная загрузка контрола 1`] = `
<div>
  <div>
    TestControlSync
  </div>
</div>
`;

exports[`UI/Async:Async в браузере Синхронная загрузка контрола, с ошибкой 1`] = `
<div>
  <div>
    У СБИС возникла проблема
  </div>
</div>
`;
